src/cloud-foundry-api-client.js

```diff
@@ -57,69 +47,85 @@ class CloudFoundryAPIClient {
+    return this._authRequest('POST', `/v3/apps/${appGUID}/tasks`, task);
+  }
+
+  stopTask(taskGUID) {
+    return this._authRequest('POST', `/v3/tasks/${taskGUID}/actions/cancel`);
+  }
+
+  getBuildContainersState(buildContainerBaseName, numBuildContainers) {
     const containerErrors = [];
-    let numBuildContainers;
-    let startedContainers;
+    let _numBuildContainers;
+    let _startedContainers;
 
-    return this.fetchBuildContainers()
+    return this.fetchBuildContainers(buildContainerBaseName, numBuildContainers)
       .then((buildContainers) => {
-        numBuildContainers = buildContainers.length;
-        startedContainers = buildContainers.filter(bc => bc.state === STATE_STARTED);
+        _numBuildContainers = buildContainers.length;
+        _startedContainers = buildContainers.filter(bc => bc.state === STATE_STARTED);
 
-        if (numBuildContainers < this._numBuildContainers()) {
-          containerErrors.push(`Expected ${this._numBuildContainers()} build containers but only ${numBuildContainers} found.`);
+        if (_numBuildContainers < numBuildContainers) {
+          containerErrors.push(`Expected ${numBuildContainers} build containers but only ${_numBuildContainers} found.`);
         }
 
-        if (startedContainers.length !== this._numBuildContainers()) {
+        if (_startedContainers.length !== numBuildContainers) {
           containerErrors.push(`Not all build containers are in the ${STATE_STARTED} state.`);
         }
 
-        return this.fetchAllAppInstanceErrors(startedContainers);
+        return this.fetchAllAppInstanceErrors(_startedContainers);
       })
       .then(instanceErrors => containerErrors.concat(instanceErrors))
       .then((errors) => {
         if (errors.length) {
           return { error: errors.join('\n') };
         }
         return {
-          expected: this._numBuildContainers(),
-          found: numBuildContainers,
-          started: startedContainers.length,
+          expected: numBuildContainers,
+          found: _numBuildContainers,
+          started: _startedContainers.length,
         };
       });
   }
 
   updateBuildContainer(container, environment) {
-    return this._authClient.accessToken()
-      .then(token => this._request(
-        'PUT',
-        container.url,
-        token,
-        { environment_json: environment }
-      ))
-      .then(() => this._authClient.accessToken())
-      .then(token => this._request(
-        'POST',
-        `${container.url}/restage`,
-        token
-      ));
-  }
-
-  _buildContainerNames() {
-    if (this._numBuildContainers() <= 1) {
-      return [this._buildContainerBaseName()];
+    return this._authRequest(
+      'PUT',
+      container.url,
+      { environment_json: environment }
+    )
+      .then(() => this._authRequest('POST', `${container.url}/restage`));
+  }
+
+  _buildContainerNames(buildContainerBaseName, numBuildContainers) {
+    if (numBuildContainers <= 1) {
```

Comment:
if numbBuildContainers is 0 should we return non-empty array?

Reply:
My idea here was that we probably won't specify the number of containers unless it is more than 1 so just assume we want 1.